basic/transit/src/lib.rs

```diff
@@ -0,0 +1,144 @@
+        #[auth(european_host_badge)]
+        pub fn withdraw_euros(&mut self, availability: bool) -> Bucket {
+            self.european_rides = availability;
+            self.collected_euros.take_all()
+        }
+
+        /// Accounts can buy tickets with dollars and euros
+        pub fn buy_ticket(&self, payment: Bucket) -> (Bucket, Bucket) {
+
+            let dollars = payment.resource_address() == self.collected_dollars.resource_address();
+            let euros = payment.resource_address() == self.collected_euros.resource_address();
+
+            let valid_ticket_price = payment.amount() >= self.ticket_price;
+            let valid_currency = dollars || euros;
+
+            scrypto_assert!(valid_ticket_price, "Invalid ticket price");
+            scrypto_assert!(valid_currency, "Invalid currency");
+
+            // Mint ticket token
+            let ticket = self.ticket_minter.authorize(|badge| {
+                self.ticket_resource_def.mint(1, badge)
+            });
+
+            // Put payment into designated collection vault
+            if dollars {
+                self.collected_dollars.put(payment.take(self.ticket_price));
+            } else {
+                self.collected_euros.put(payment.take(self.ticket_price));
+            }
+
+            // Return minted ticket along with payment change (e.g. ticket=10, payment=15, change=5)
+            (ticket, payment)
+        }
+
+        /// Accounts can go on rides with tickets they bought (minted)
+        pub fn ride(&mut self, payment: Bucket, ride_type: String) {
+
+            let valid_ride = (ride_type.eq("American") && self.american_rides) || (ride_type.eq("European") && self.european_rides);
+            let valid_currency = payment.resource_address() == self.ticket_resource_def.address();
+            let valid_price_per_ride = payment.amount() == self.ride_price;
+
+            scrypto_assert!(valid_ride, "Invalid ride");
+            scrypto_assert!(valid_currency, "Invalid currency");
+            scrypto_assert!(valid_price_per_ride, "Invalid price per ride");
+
+            // Burn ticket token
+            self.ticket_minter.authorize(|badge| {
+                payment.burn(badge);
+            });
+
+            // Keep track of riders public key/epoch
+            if self.riders.get(&Context::transaction_signers()) == None {
+                info!("Hi, this is your first ride on a transit, have fun!");
+                self.riders.insert(Context::transaction_signers(),Context::current_epoch());
+            } else if self.riders.get(&Context::transaction_signers()).unwrap() == Context::current_epoch() {
+                info!("Hi, you have already used the transit more than once during epoch: {}", Context::current_epoch())
+            } else {
+                info!("Hi, welcome back, you have not used the transit during epoch: {}", Context::current_epoch());
+                self.riders.insert(Context::transaction_signers(),Context::current_epoch());
```

Comment:
This would be perfect candidate for non-fungible NFT. You issue an NFT for each rider and record the epoch time inside the NFTs.